[ready] Add logdet and slogdet

ssnl opened the pull request:
See commit messages for details of each commit. Notice that the last commit changes the QR based approach to LU based.

Some comments about switching from QR to LU:

1. QR (`geqrf`) is generally more stable than LU (`getrf`).
2. But `getrf` gives specific information about matrix being singular or not. Furthermore, the backward (when det > 0) uses matrix inverse (`getrf` + `getri`), so using `getrf` in forward makes sense.
3. Finally, this is also more consistent with many other libraries (including NumPy and MATLAB) that use LU factorization (`getrf`) to compute determinant. 

Therefore, in this PR, det is computed using `btrtifact` (batched `getrf`). However, the backward still uses `inverse` (`getrf` + `getri`). So future work include:
1. expose `getri` and cache `getrf` results to speed up backward.
2. implement batched `getri` to support batched det methods.

SsnL commented:
@pytorchbot retest this please

ssnl committed:
1. Add logdet and slogdet in ATen side
2. Previously, det can return result with incorrect sign upon seeing symmetric
   matrices. This is caused by the wrong assumption I had on SVD (when input is
   symmetric U=V^T). This fixes it.
3. Moreover, after fixing 2 now QR is always needed for det forward. So I moved
   SVD to backward call. Since this is a specific variant of SVD, it is named as
   _svd_with_positive_UV_det, with derivative.yaml entry being svd_backward.
4. Updated/added backward functions for det, logdet and slogdet, which uses
   _svd_with_positive_UV_det and svd_backward inside.
5. Optimized svd_backward:
   a. Avoid unnecessary kernels when only sigma has gradient (this is the usual
      case, and also true with *det backward functions).
   b. Fix SVD double backward by avoiding a nan.

ssnl committed:
1. Add/update grad checks for det, logdet, and slogdet.
2. Fix an incorrect check for dim_args_idx in test_autograd.py
3. Add option to only test a subset of output values, specified by
   test_output_indices, for cases like slogdet where only the
   second output is differentiable.
4. Add better doc for the test generating list.

ssnl committed:
Add/improve output tests for det, logdet and slogdet
Add a scaling to random matrices so closeness checks are more robust

ssnl committed:
Remove unnecessaery Variable wrappers in some test files

ssnl committed:
Add logdet slogdet docs

ssnl committed:
Improve an err msg in THTensorLapack.c

vishwakftw commented:
@SsnL I will have to use logdet for the implementation of the KL-divergence between two multivariate normal distributions. I would like to know how fast would this be as opposed to computing the determinant and taking the log of it?

soumith commented:
putting speed aside, the gradient formulas permitting this would prob be more numerically stable

SsnL commented:
@pytorchbot retest this please

SsnL commented:
@pytorchbot retest this please

SsnL commented:
@pytorchbot retest this please

SsnL commented:
@pytorchbot retest this please

SsnL commented:
@pytorchbot retest this please

SsnL commented:
@zou3519 can you review this when you have time? I moved some doc changes to other PR so it's shorter now. :) Basically, maths is:

In forward: use QR factorization to get A=QR, then det(A) = det(Q) det(R). det(Q) is obtained by counting the number on nnz entries in tau (-1 if odd, and 1 if even). And det(R) is the product of diag(R).

In backward: calculate a specific variant of SVD: A = USV^T, s.t. det(UV) = 1. So det(A) = det(S) = prod(diag(S)). Then use svd_backward to backprop the determinant as if it is calculated using this method.

SsnL commented on test/test_autograd.py:
```diff
@@ -2532,13 +2566,34 @@ class dont_convert(tuple):
+     'symmetric_pd', NO_ARGS, [skipIfNoLapack]),
+    ('logdet', lambda: make_nonzero_det(random_fullrank_matrix_distinct_singular_value(S), 1, 0), NO_ARGS,
+     'distinct_singular_values', NO_ARGS, [skipIfNoLapack]),
+    ('slogdet', lambda: make_nonzero_det(torch.randn(S, S), 1), NO_ARGS, 'pos_det', NO_ARGS, [skipIfNoLapack], [1]),
```
the last element of [1] means that only the 2nd output of slogdet is differentiable, and should be gradcheck'ed

ssnl committed:
add inverse-based backward for invertible matrices
use svd only for non-invertible case, so don't need the special variant anymore

zou3519 commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -8,50 +8,100 @@
+//
 // http://www.ics.forth.gr/cvrl/publications/conferences/2000_eccv_SVD_jacobian.pdf
-// But instead of gesvd SVD A = U(A) Sig(A) V(A)^T, which doesn't specify signs
+// Instead of gesvd SVD A = U(A) Sig(A) V(A)^T, which doesn't specify signs
```
What do you mean by "doesn't specify signs?"

SsnL replied:
signs of determinants of U and V. The sentence continues on the next line :). No worries. I'll ask someone with more experience to review this.

SsnL commented:
Hold on reviewing @zou3519 . I plan to move det computation from QR to LU factorization. It would be more consistent with things like numpy matlab and also won't be having problem with the inverse in backward.

ssnl committed:
use LU rather than QR

SsnL commented:
Some comments about switching to LU:

1. QR (`geqrf`) is generally more stable than LU (`getrf`).
2. But `getrf` gives specific information about matrix being singular or not. Furthermore, the backward (when det > 0) uses matrix inverse (`getrf` + `getri`), so using `getrf` in forward makes sense.
3. Finally, this is also more consistent with many other libraries (including NumPy and MATLAB) that use LU factorization (`getrf`) to compute determinant. 

Therefore, in this PR, det is computed using `btrtifact` (batched `getrf`). However, the backward still uses `inverse` (`getrf` + `getri`). So future work include:
1. expose `getri` and cache `getrf` results to speed up backward.
2. implement batched `getri` to support batched det methods.

SsnL commented:
@vishwakftw @samuela  do one of you want to review this PR? Thanks in advance! :)

vishwakftw reviewed (approved):
I think the code looks good. Thanks!!

SsnL commented:
@vishwakftw @soumith Thanks!!!
